File: leetcode/solutions/0599.MinimumIndexSumOfTwoLists/findRestaurant.go
package solution

func findRestaurant(list1 []string, list2 []string) []string {
	nameIndex := make(map[string]int, len(list1))

	for i, name := range list1 {
		nameIndex[name] = i
	}

	// init with large enough number
	minSum := len(list1) + len(list2)
	result := make([]string, 0, len(list1)+len(list2))

	for i, name := range list2 {
		if j, ok := nameIndex[name]; ok {
			if i+j < minSum {
				result = []string{name}
				minSum = i + j
			} else if i+j == minSum {
				result = append(result, name)
			}
		}
	}

	return result
}

File: leetcode/solutions/0599.MinimumIndexSumOfTwoLists/findRestaurant.ts
export default function findRestaurant(
  list1: string[],
  list2: string[]
): string[] {
  const map = new Map<string, number>();

  for (const [i, name] of list1.entries()) {
    map.set(name, i);
  }

  let minSum = Infinity;
  let result: string[] = [];

  for (const [j, name] of list2.entries()) {
    const i = map.get(name) ?? Infinity;

    if (i + j < minSum) {
      result = [name];
      minSum = i + j;
    } else if (i + j === minSum) {
      result.push(name);
    }
  }

  return result;
}
